File: 03 Blind Search/blind.R
# metodo de busqueda ciega
# 	objetivo - matriz con las soluciones x D
# 	Fx - funcion a evaluar
# 	type - "min" o "max"
# 	... - parametros extras para Fx

fsearch = function(objetivo, Fx, type="min", ...)
{
	x = apply(objetivo,1,Fx,...) # evalua Fx sobre todas las filas
	ib = switch(type, min=which.min(x), max=which.max(x))
	return (list(index=ib, sol=objetivo[ib,], eval=x[ib]))
}

# metodo de busqueda completa en profundidad
# 	l - nivel del arbol
# 	b - rama del arbol
# 	dominio - lista de vectores de tamaño D con los valores de dominio
# 	Fx - funcion a evaluar
# 	type - "min" o "max"
# 	D - dimension (numero de variables)
# 	x - vector de soluciones actual
# 	msol - mejor solucion actual
# 	... - parametros extras para Fx

dfsearch = function(l=1, b=1, dominio, Fx, type="min", D=length(dominio),
		x=rep(NA, D), msol=switch(type, min=list(sol=NULL, eval=Inf),
									max=list(sol=NULL, eval=-Inf)), ... )
{
	if ((l-1)==D) # evalua la rama superior
	{
		f = Fx(x,...)
		fb = msol$eval
		ib = switch(type, min=which.min(c(fb, f)),max=which.max(c(fb, f)))
		if (ib==1) return (msol) else return(list(sol=x, eval=f))
	}
	else # pasa por las sub ramas
	{
		for (j in 1:length(dominio[[l]]))
		{
			x[1] = dominio[[l]][j]
			msol = dfsearch(l+1, j, dominio, Fx, type, D=D, x=x, msol=msol, ... )
		}
		return(msol)
	}
}

mcsearch = function(N,lower,upper,FUN,type="min",...) 
{ 
	D=length(lower) 
	s=matrix(nrow=N,ncol=D) # set the search space 
	for(i in 1:N) s[i,]=runif(D,lower,upper) 
	fsearch(s,FUN,type,...) # best solution
}

N= 1000

sphere=function(x) sum(x^2)
rastrigin=function(x) 10*length(x)+sum(x^2-10*cos(2*pi*x))
D=c(2,30)

label="sphere" 
for(i in 1:length(D)) 
{ 
	S=mcsearch(N,rep(-5.2,D[i]),rep(5.2,D[i]),sphere,"min") 
	cat(label,"D:",D[i],"s:",S$sol[1:2],"f:",S$eval,"\n") 
}
label="rastrigin" 
for(i in 1:length(D)) 
{ 
	S=mcsearch(N,rep(-5.2,D[i]),rep(5.2,D[i]),rastrigin,"min") 
	cat(label,"D:",D[i],"s:",S$sol[1:2],"f:",S$eval,"\n") 
}

File: 03 Blind Search/blind.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-

import pandas as pd
import numpy as np 
import math

def fsearch(Fx, sol, type="min"):
	"""Método de busqueda ciega
	Halla la solucion de la función Fx en el espacio de busqueda.
	Argumentos
	----------
	Fx - Función
	sol - Espacio de busqueda, Matrix(MxN)
	type (opcional) - Tipo de optimización

	Devuelve
	--------
	x - Solucion de la funcion Fx en el espacio de busqueda
	"""
	if (not type == "min" and not type == "max"):
		raise ValueError("Metodo de optimización no valido")

	x = pd.DataFrame((sol))
	y = x.apply(Fx,1)
	ib = np.min(y) if type == "min" else np.max(y)
	return ib[0]

def dfsearch(dominio, Fx, l=0, b=1, type="min", D=None, x=None, msol=None):
	"""Método de busqueda ciega completa 
	Halla la solucion de la función Fx en el espacio de busqueda.
	Argumentos
	----------
	Fx - Función
	dominio - Espacio de busqueda, Matrix(MxN)
	l - Nivel del arbol
	b - Rama del arbol
	type (opcional) - Tipo de optimización
	D - Dimensión (numero de variables)
	x - vector de posibles soluciones
	msol - mejor solucion actual 
	
	Devuelve
	--------
	x - Solucion de la funcion Fx en el espacio de busqueda
	"""
	if (not type == "min" and not type == "max"):
		raise ValueError("Metodo de optimización no valido")

	D = len(dominio) if D == None else D
	x = [None]*len(dominio) if x== None else x
	msol = ([None, float('inf')]if type=="min" else [None, float('-inf')]) if msol == None else msol
	if ((l-1)==D):
		f = pd.DataFrame((dominio))
		f = f.apply(Fx)
		fb = msol[1]
		ib = np.min(f) if type == "min" else np.max(f)
		ib = sorted(ib, reverse=True)[0]
		return np.min(msol[0]) if ib==1 else x
	else:
		for j in range(len(dominio)):
			x[l-1]= dominio[l-1][j]
			msol = dfsearch(dominio, Fx, l+1, j, type, x=x, msol=msol)
		return msol

def f(x):
	return pow(x,3)+ 4*pow(x,2)-10
	#return pow((x-7),3)+ pow((2*x-5),2)

def espacio(m, n, r1, r2):
	dat = np.zeros((m,n))
	for i in range(m):
		for j in range(n):
			dat[i][j] = np.random.randint(r1, r2)
	return dat 

a = 0	#limite inferior
b = 9	#limite superior
c = 3	#numero de variables
d = 5	#tamaño del espacio
metodo= dfsearch(espacio(c,d,a,b),f, type="min")
print(sorted(metodo))
print("x: ",sorted(metodo, reverse=True))
print("y: ",f(sorted(metodo, reverse=True)[0]))
